### PR TITLE
feat: new deposit registry algorithm

### PR DESCRIPTION
**Pull request by eddort**

### CODE REVIEW

File: src/contracts/deposits-registry/deposits-registry.service.ts [hunk 1]
```diff
@@ -159,6 +159,10 @@ export class DepositRegistryService {
         finalizedBlock,
         finalizedBlockHash,
```
[the remark this answers is not in the document]
Comment (Amuhar) [reply src/contracts/deposits-registry/deposits-registry.service.ts:160]: why also we decided not to check that data is valid in cache and there are no events with depositCount > lastValidEvent ?

File: src/contracts/deposits-registry/sanity-checker/integrity-checker/deposit-tree/deposit-tree.ts [hunk 2]
```diff
@@ -68,12 +68,20 @@ export class DepositTree {
...
+   * Inserts a new node into the tree using an already computed hash. The insertion only proceeds
+   * if the deposit count provided is the next sequential number expected (one more than the current node count).
+   * @param {Uint8Array} node - The hash of the node to be inserted, represented as a Uint8Array.
+   * @param {bigint} depositCount - The sequential count of the deposit event from the blockchain,
+   *                                expected to be one more than the current highest node count.
+   * @returns {boolean} Returns true if the node was successfully inserted, false otherwise.
    */
-  public insert(node: Uint8Array) {
+  public insert(node: Uint8Array, depositCount: bigint): boolean {
+    if (depositCount !== this.nodeCount + 1n) {
```
Comment (Amuhar) [inline src/contracts/deposits-registry/sanity-checker/integrity-checker/deposit-tree/deposit-tree.ts:79]: good finding 👍

File: src/contracts/deposits-registry/store/store.service.spec.ts [hunk 3]
```diff
@@ -4,7 +4,15 @@ import { ConfigModule } from 'common/config';
...
 import { DepositsRegistryStoreModule } from './store.module';
 import { DepositsRegistryStoreService } from './store.service';
-import { cacheMock } from './store.fixtures';
+import { cacheMock, eventMock1 } from './store.fixtures';
+
+const getEventsDepositCount = async (
+  dbService: DepositsRegistryStoreService,
+) => {
+  const result = await dbService.getEventsCache();
+  const remainingDeposits = result.data.map((event) => event.depositCount);
```
Comment (Amuhar) [inline src/contracts/deposits-registry/store/store.service.spec.ts:13]: why Remaining ?

File: src/contracts/deposits-registry/store/store.service.spec.ts [hunk 4]
```diff
@@ -48,4 +56,79 @@ describe('dbService', () => {
...
   });
+
+  describe('deleteDepositsGreaterThanNBatch', () => {
+    const testCases = [
+      { N: 10, deposits: [9, 10, 11, 12], expectedRemaining: [9, 10] },
+      { N: 5, deposits: [3, 4, 5, 6], expectedRemaining: [3, 4, 5] },
+      { N: 0, deposits: [0, 1, 2], expectedRemaining: [0] },
+    ];
+
+    it.each(testCases)(
```
Comment (Amuhar) [inline src/contracts/deposits-registry/store/store.service.spec.ts:67]: good idea 👍

File: src/contracts/deposits-registry/store/store.service.spec.ts (hunk 3 above)
Comment (eddort) [reply src/contracts/deposits-registry/store/store.service.spec.ts]: fixed

File: src/contracts/deposits-registry/deposits-registry.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddort) [reply src/contracts/deposits-registry/deposits-registry.service.ts:160]: I've added such a check elsewhere, where it's easier to check and gives more security

[the remark this answers is not in the document]
Comment (eddort) [reply src/contracts/deposits-registry/deposits-registry.service.ts:160]: > why dont throw error at 91 raw ?
> 
> ```ts
> public async updateEventsCache(): Promise<void> {
>     ...
>     if (!isCacheValid) return; 
> ```

in this case it is not necessary to throw an exception

[the remark this answers is not in the document]
Comment (Amuhar) [reply src/contracts/deposits-registry/deposits-registry.service.ts:160]: why ?

File: test/duplicates-v3.e2e-spec.ts [hunk 5]
```diff
@@ -100,6 +100,9 @@ describe('Deposits in case of duplicates', () => {
       .mockImplementation(() => Promise.resolve());
 
     // deposit cache mocks
+    jest
```
Comment (Amuhar) [inline test/duplicates-v3.e2e-spec.ts:103]: What was changed in the PR to make this mock necessary?

Comment (Amuhar) [reply test/duplicates-v3.e2e-spec.ts:103]: before commit 571dc751094e2f7f0734c1204203d2307f084712 tests worked

Comment (eddort) [reply test/duplicates-v3.e2e-spec.ts:103]: because of the index check before inserting the event to the deposit tree

**Review by Amuhar — APPROVED** [review]

**Pull request merged**